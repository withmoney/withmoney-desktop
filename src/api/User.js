import axios from 'axios';

export const login = data => axios.post('http://localhost:3000/api/v1/login', data);

export const signup = data => axios.post('http://localhost:3000/api/v1/signup', data);
