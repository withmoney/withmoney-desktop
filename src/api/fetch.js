import axios from 'axios';
import { catchInvalidToken } from 'app/utils/parse';

const newAxios = axios.create({
  baseURL: `${process.env.MYMONEY_API}`,
  headers: {
    Authorization: {
      toString() {
        return `Bearer ${localStorage.getItem('token')}`;
      },
    },
  },
});

const fetch = method => (...args) => catchInvalidToken(newAxios[method](...args));

export default {
  get: fetch('get'),
  post: fetch('post'),
  put: fetch('put'),
  delete: fetch('delete'),
};
