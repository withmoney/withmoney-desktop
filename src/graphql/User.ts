import { gql } from '@apollo/client';

export const UPDATE_USER = gql`
  mutation updateUser($firstName: String!, $lastName: String!, $email: String!, $language: Locale) {
    updateOneUser(
      data: { firstName: $firstName, lastName: $lastName, email: $email, language: $language }
    ) {
      id
      firstName
      lastName
      email
      language
    }
  }
`;
