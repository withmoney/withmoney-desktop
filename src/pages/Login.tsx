import React, { FormEvent, useState, ChangeEvent } from 'react';
import styled from 'styled-components';
import Auth from '../services/auth';

const Page = styled.div`
  height: 100vh;
  display: flex;
  flex-direction: column;
  background-color: #f2f2f2;
`;

const Container = styled.div`
  padding-left: 15px;
  padding-right: 15px;
`;

const Brand = styled.h1`
  text-align: center;
  font-size: 42px;
  font-weight: 700;
  margin-top: 70px;
  margin-bottom: 45px;
`;

const Form = styled.form`
  background-color: white;
  margin-left: auto;
  margin-right: auto;
  max-width: 445px;
  width: 100%;
  border-radius: 5px;
  padding: 20px 20px 30px;
  display: flex;
  flex-direction: column;
`;

const Title = styled.h2`
  font-size: 32px;
  font-weight: 700;
  text-align: center;
  margin-bottom: 20px;
  flex-direction: column;
`;

const Field = styled.input`
  font-size: 16px;
  border-radius: 5px;
  border: 1px solid #bdbdbd;
  padding: 10px 12px;
  margin-bottom: 20px;
  outline: none;

  &:focus {
    box-shadow: 0 0 0px 2px rgba(33, 150, 83, 0.5);
  }
  &:disabled {
    color: gray;
  }
`;

const Flex = styled.div`
  display: flex;
  justify-content: space-between;
  & + & {
    margin-top: 20px;
  }
`;

const Button = styled.button`
  background-color: #219653;
  color: white;
  font-size: 16px;
  border-width: 0px;
  padding: 11px 13px;
  border-radius: 5px;
  cursor: pointer;
  outline: none;

  &:hover {
    background-color: #308856;
  }

  &:focus {
    box-shadow: 0 0 0px 2px rgba(33, 150, 83, 0.5);
  }

  &:active {
    background-color: #256a43;
  }

  &:disabled {
    background-color: hsla(146, 10%, 36%, 1);
  }
`;

const Link = styled.a`
  color: #219653;
  text-decoration: underline;

  &:hover {
    color: #308856;
  }
  &:active {
    color: #256a43;
  }
`;

interface AlertProps {
  isRed?: boolean;
}

const Alert = styled.div`
  background-color: ${(props: AlertProps) => (props.isRed ? '#963621' : '#214296')};
  color: white;
  padding: 20px;
  margin-bottom: 20px;
  border-radius: 4px;
`;

const Login = () => {
  const [form, setForm] = useState({ email: '', password: '' });
  const [isLoading, setIsLoading] = useState(false);
  const [error, setError] = useState('');

  const onSubmit = async (event: FormEvent<HTMLFormElement>) => {
    event.preventDefault();

    setIsLoading(true);
    try {
      const { data } = await Auth.login(form);
      if (!data.success) {
        setError('Your email or password are invalid!');
      }
    } catch (error) {
      setError(error.message);
    } finally {
      setIsLoading(false);
    }
  };

  const handleInput = (event: ChangeEvent<HTMLInputElement>) => {
    const { name, value } = event.target;
    setForm({
      ...form,
      [name]: value
    });
  };

  return (
    <Page>
      <Container>
        <Brand>withmoney</Brand>
        <Form onSubmit={onSubmit}>
          <Title>Signin</Title>
          {!!error && <Alert isRed>{error}</Alert>}
          <Field
            type="email"
            name="email"
            placeholder="Email"
            onChange={handleInput}
            disabled={isLoading}
            required
          />
          <Field
            type="password"
            name="password"
            placeholder="Password"
            onChange={handleInput}
            disabled={isLoading}
            required
          />
          <Flex>
            <Link>Reset your password</Link>
            <Button disabled={isLoading}>{isLoading ? 'Sending...' : 'Login'}</Button>
          </Flex>
          <Flex>
            <span>Do you not have an account?</span>
            <Link>Signup</Link>
          </Flex>
        </Form>
      </Container>
    </Page>
  );
};

export default Login;
