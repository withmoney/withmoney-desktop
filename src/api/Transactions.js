import fetch from 'api/fetch';
import { mountQuery } from 'app/utils/parse';

export const list = (query = {}) => fetch.get(mountQuery('transactions', query));

export const get = id => fetch.get(`transactions/${id}`);

export const put = (id, data) => fetch.put(`transactions/${id}`, data);
