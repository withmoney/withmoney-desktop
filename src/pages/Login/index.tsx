import React, { FormEvent, useState, ChangeEvent, useEffect } from 'react';
import { useMutation } from '@apollo/react-hooks';
import { useHistory } from 'react-router-dom';
import { toast } from 'react-toastify';
import { USER_LOGIN } from '../../graphql/AuthGql';
import { loginSchema } from '../../schema/auth';
import Button from '../../components/Button';
import Input from '../../components/Input';
import Page from '../../components/Page';
import Header from '../../components/Header';
import Form from '../../components/Form';
import Flex from '../../components/Flex';
import Link from '../../components/Link';
import Container from '../../components/Container';
import Text from '../../components/Text';
import InputGroup from '../../components/InputGroup';
import InputControl from '../../components/InputControl';

const initialValues = {
  email: '',
  password: '',
};

const Login = () => {
  const [form, setForm] = useState({ email: '', password: '' });
  const [formState, setFormState] = useState({ error: initialValues, isValid: false });
  const [userLogin, { loading }] = useMutation(USER_LOGIN);
  const history = useHistory();

  useEffect(() => {
    const checkForm = async () => {
      try {
        const isValid = await loginSchema.isValid(form);
        setFormState({ error: { ...initialValues }, isValid: isValid });
      } catch (err) {
        setFormState({ error: { ...initialValues }, isValid: false });
      }
    };
    checkForm();
  }, [form]);

  const handleInput = async (event: ChangeEvent<HTMLInputElement>) => {
    const { name, value } = event.target;
    setForm({
      ...form,
      [name]: value.trim(),
    });
  };

  const handleBlur = async (event: ChangeEvent<HTMLInputElement>) => {
    const { name } = event.target;
    try {
      await loginSchema.validateAt(name, form);
      const isValid = await loginSchema.isValid(form);
      setFormState({ error: { ...formState.error, [name]: '' }, isValid: isValid });
    } catch (err) {
      setFormState({ error: { ...formState.error, [name]: err.message }, isValid: false });
    }
  };

  const onSubmit = async (event: FormEvent<HTMLFormElement>) => {
    event.preventDefault();
    if (await loginSchema.isValid(form)) {
      try {
        const {
          data: {
            login: { token },
          },
        } = await userLogin({
          variables: form,
        });

        localStorage.setItem('withmoney-token', token);
        history.push('/');
      } catch (err) {
        toast.error(err.message);
      }
    }
  };

  return (
    <Page>
      <Container>
        <Header as="h1" align="center">
          withmoney
        </Header>

        <Form onSubmit={onSubmit}>
          <Header as="h3" align="center">
            Log in
          </Header>

          <InputControl message={formState.error.email} isInvalid={!!formState.error.email}>
            <InputGroup>
              <Input
                isInvalid={!!formState.error.email}
                type="email"
                name="email"
                placeholder="Email"
                disabled={loading}
                onBlur={handleBlur}
                onChange={handleInput}
              />
            </InputGroup>
          </InputControl>

          <InputControl message={formState.error.password} isInvalid={!!formState.error.password}>
            <InputGroup>
              <Input
                isInvalid={!!formState.error.password}
                type="password"
                name="password"
                placeholder="Password"
                disabled={loading}
                onChange={handleInput}
                onBlur={handleBlur}
              />
            </InputGroup>
          </InputControl>

          <Flex justifyContent="space-between">
            <Link to="/change-password" variation="primary">
              Reset your password
            </Link>

            <Button variation="primary" disabled={!formState.isValid || loading}>
              {loading ? 'Sending...' : 'Log in'}
            </Button>
          </Flex>

          <Flex justifyContent="space-between">
            <Text>Do you not have an account?</Text>
            <Link to="/signup" variation="primary">
              Sign up
            </Link>
          </Flex>
        </Form>
      </Container>
    </Page>
  );
};

export default Login;
