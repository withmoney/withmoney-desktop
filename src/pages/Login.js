import React, { Fragment } from 'react';
import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';
import * as UserApi from '../api/User';
import * as UserAction from '../store/modules/user';
import BoxForm from '../components/BoxForm';
import FieldInput from '../components/FieldInput';
import FieldButton from '../components/FieldButton';

console.log(UserAction);

class Login extends React.Component {
  constructor(props) {
    super(props);

    this.handleChange = this.handleChange.bind(this);
    this.onSave = this.onSave.bind(this);

    this.state = {
      email: '',
      password: '',
    };
  }

  async onSave(event) {
    event.preventDefault();
    const { actions } = this.props;
    try {

      const { data } = await UserApi.login(this.state);

      if (data.success) {
        window.localStorage.setItem('token', data.token);
        window.localStorage.setItem('user', JSON.stringify(data.payload));

        actions.setUser({
          token: data.token,
          data: data.payload,
        });
      } else {
        console.log('User not finded.');
      }
    } catch (e) {
      console.log('User not finded.');
    }
  }

  handleChange({ target }) {
    this.setState({
      [target.id]: target.value,
    });
  }


  render() {
    console.log(this.props)
    return (
      <BoxForm
        title="withmoney"
        subtitle="Log in"
        onSubmit={this.onSave}
        fields={(
          <Fragment>
            <FieldInput id="email" onChange={this.handleChange} />
            <FieldInput id="password" type="password" onChange={this.handleChange} />
            <FieldButton type="submit">
              Log in
            </FieldButton>
          </Fragment>
        )}
        footer={(
          <Fragment>
            <span>Do not have an account?</span>
            <a href="#">Signup</a>
          </Fragment>
        )}
      />
    );
  }
}

const mapStateToProps = ({ user }) => ({ user });

const mapDispatchToProps = dispatch => ({
  actions: bindActionCreators(UserAction, dispatch),
});

export default connect(mapStateToProps, mapDispatchToProps)(Login);
