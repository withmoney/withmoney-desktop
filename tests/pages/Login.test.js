import renderer from 'react-test-renderer';
import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import { mount } from 'enzyme';
import Login from '../../src/pages/Login';
import * as UserApi from '../../src/api/User';

const mockStore = configureStore([thunk]);

jest.mock('../../src/api/User', () => ({
  login: jest.fn().mockResolvedValue({
    data: {
      success: true,
    },
  }),
}));

describe('Login', () => {
  let Component;

  beforeAll(() => {
    Component = global.withReduxAndRouter(mockStore({}), Login);
  });

  it('should render a component correctly', () => {
    const wrapper = renderer.create(Component());

    expect(wrapper.toJSON()).toMatchSnapshot();
  });

  it('should call onSabe action when login form is submitted', () => {
    const wrapper = mount(Component());

    wrapper
      .find('input[name="email"]')
      .simulate('change', { target: { name: 'email', value: 'davidcostadev@gmail.com' } });
    wrapper
      .find('input[name="password"]')
      .simulate('change', { target: { name: 'password', value: '123456' } });

    wrapper.find('form').simulate('submit');

    expect(UserApi.login).toBeCalledWith({
      email: 'davidcostadev@gmail.com',
      password: '123456',
    });
  });
});
