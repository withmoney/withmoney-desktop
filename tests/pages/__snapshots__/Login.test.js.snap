// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login should render a component correctly 1`] = `
<div
  className="box-form"
>
  <h1
    className="box-form__title"
  >
    withmoney
  </h1>
  <form
    className="box-form__form"
    onSubmit={[Function]}
  >
    <h3
      className="box-form__subtitle"
    >
      Log in
    </h3>
    <div
      className="field"
    >
      <input
        className="field__input"
        id="email"
        name="email"
        onChange={[Function]}
        type="text"
      />
    </div>
    <div
      className="field"
    >
      <input
        className="field__input"
        id="password"
        name="password"
        onChange={[Function]}
        type="password"
      />
    </div>
    <div
      className="field"
    >
      <button
        className="button"
        type="submit"
      >
        Log in
      </button>
    </div>
    <div
      className="box-form__footer"
    >
      <span>
        Do not have an account?
      </span>
      <a
        href="/signup"
        onClick={[Function]}
      >
        Sign Up
      </a>
    </div>
  </form>
</div>
`;
